src/JSON.jl

```diff
@@ -61,221 +74,253 @@ lower(d::DataType) = string(d)
-        printsp(io, state)
-        _writejson(io, state, x)
-    end
-    end_object(io, state, false)
+    end_object(io)
 end
 
-function _writejson(io::IO, state::State, a)
-    # FIXME: This fallback is harming performance substantially.
-    # Remove this fallback when _print removed.
-    if applicable(_print, io, state, a)
-        Base.depwarn(
-            "Overloads to `_print` are deprecated; extend `lower` instead.",
-            :_print)
-        _print(io, state, a)
-    else
-        _writejson(io, state, lower(a))
+function show_json(io::SC, s::CS, x::Union{AbstractVector, Tuple})
+    begin_array(io)
+    for elt in x
+        show_element(io, s, elt)
     end
+    end_array(io)
 end
 
-# Note: Arrays are printed in COLUMN MAJOR format.
-# i.e. json([1 2 3; 4 5 6]) == "[[1,4],[2,5],[3,6]]"
-function _writejson{T, N}(io::IO, state::State, a::AbstractArray{T, N})
-    lengthN = size(a, N)
-    if lengthN > 0
-        start_object(io, state, false)
-        if VERSION <= v"0.3"
-            newdims = ntuple(N - 1, i -> 1:size(a, i))
-        else
-            newdims = ntuple(i -> 1:size(a, i), N - 1)
-        end
-        Base.print(io, prefix(state))
-        _writejson(io, state, Compat.view(a, newdims..., 1))
-
-        for j in 2:lengthN
-            Base.print(io, ",")
-            printsp(io, state)
-            _writejson(io, state, Compat.view(a, newdims..., j))
-        end
-        end_object(io, state, false)
-    else
-        Base.print(io, "[]")
+"""
+Serialize a multidimensional array to JSON in column-major format. That is,
+`json([1 2 3; 4 5 6]) == "[[1,4],[2,5],[3,6]]"`.
+"""
+function show_json{T,n}(io::SC, s::CS, A::AbstractArray{T,n})
+    begin_array(io)
+    newdims = ntuple(_ -> :, Val{n - 1})
+    for j in 1:size(A, n)
+        show_element(io, s, Compat.view(A, newdims..., j))
     end
+    end_array(io)
 end
 
-# this is _print() instead of _print because we need to support v0.3
-# FIXME: drop the parentheses when v0.3 support dropped
-"Deprecated way to overload JSON printing behaviour. Use `lower` instead."
-function _print(io::IO, s::State, a::JSONPrimitive)
-    Base.depwarn(
-        "Do not call internal function `JSON._print`; use `JSON.print`",
-        :_print)
-    _writejson(io, s, a)
-end
+show_json(io::SC, s::CS, a) = show_json(io, s, lower(a))
 
-function print(io::IO, a, indent=0)
-    _writejson(io, State(indent), a)
-    if indent > 0
-        Base.print(io, "\n")
+# Fallback show_json for non-SC types
+"""
+Serialize Julia object `obj` to IO `io` using the behaviour described by `s`. If
+`indent` is provided, then the JSON will be pretty-printed; otherwise it will be
+printed on one line. If pretty-printing is enabled, then a trailing newline will
+be printed; otherwise there will be no trailing newline.
+"""
+function show_json(io::IO, s::Serializations.Serialization, obj; indent=nothing)
+    ctx = indent === nothing ? CompactContext(io) : PrettyContext(io, indent)
```

Comment:
The previous version, using `State{true}` or `State{false}`, was type unstable. This one is actually not (in the sense that the type of `ctx` only depends on the type of `indent`, instead of its value). However, the keyword argument sorter is type unstable anyway.